### ISSUE TITLE
Add a `Float` generation method

### ISSUE THREAD

**Issue by rlouf** [normal-computing/outlines#154]
We can steer the generation process so we only output floats. I propose to implement a `Float` subclass of `Sequence` that uses [masking](https://github.com/normal-computing/outlines/blob/a4662218232df8264d1b9eb93867362ed8da2e54/outlines/text/masks.py#L36) to restrict the generated tokens to floats. The mask is a function of the tokens that have already been generated: if we sequence generated so far contains a period we can only generate integers. We will need to add a `create_proposal` method to `Sequence` that applies the mask to the logits generated by the model.

We can also add constraints on the generated floats:

* Enforce a given precision
* Add an upper or lower bound on the value of the float

We will probably need [SMC sampling](https://arxiv.org/abs/2306.03081v1) to implement the constraints.

**Comment by brandonwillard** [comment]
If we're talking about regex-driven float parsing, we can also use the vocabulary pre-processing approach described in #131 and [this Gist](https://gist.github.com/brandonwillard/4567b124596835c0dec31df7cd094da8).  That could possibly turn the process of determining valid next tokens (and/or the respective indices to be masked) into something closer to a simple `dict` look-up.

**Comment by brandonwillard** [comment]
Looks like the current float masking occasionally produces strings like `".801.4"` in `test_hf_transformers.test_type_float`: see the CI failure in a run of #172 [here](https://github.com/normal-computing/outlines/actions/runs/5479822931/jobs/9982184249#step:5:47).

The FSM-based pre-processing approach mentioned above and utilized in https://github.com/normal-computing/outlines/pull/166 should fix that.
